### ISSUE TITLE
Data type for storing datetime values and timezone handling

### ISSUE THREAD

**Issue by klesh** [merico-dev/lake#426]
## Description

Now we have two data-types for `DATETIME` storage:

1. `time.Time` / `DATETIME` like `jira plugin`
2. `string` / `longtext` like `gitlab plugin`

Both have their pros and cons.
For `DATETIME` type we have to deal with the timezone problem, because mysql doesn't store timezone information for `DATETIME` type.
For `longtext`, it's hard for user to query on Grafana Dashboard, they have to manually convert string to `DATETIME` type, do the filtering then convert it back, makes query tedious. And there is performance issue for indexing.

## Describe the solution you'd like

For the sake of user friendly, I suggest we opt for `DATETIME`, and address the timezone problem.
The solution is to store/load them with a consistence timezone, and this timezone must be what user prefers, and could be modified when needed.

### Some information I found:

> In MySQL 8.0.19 and later, you can specify a time zone offset when inserting a TIMESTAMP or DATETIME value into a table. See Section 9.1.3, “Date and Time Literals”, for more information and examples.
> https://dev.mysql.com/doc/refman/8.0/en/datetime.html

If I understand it correctly, mysql will convert date-string like `2038-01-19 08:00:00+08:00` to a specific timezone, for example, if  that specific timezone was UTC, then it became `2038-01-19 00:00:00` in database.  The specific timezone can be configured through connection string. 

> gorm配置的时区为使用本地时区，通过查询数据库表可以观察得到，生成的时间为本地当前时间。
> https://teroy.github.io/2018/12/01/gorm-loc/

According to above article,  by setting up `loc` we specify the timezone that mysql uses to store `DATETIME` values.

### How `time.Time` value changes during store/load process

I've created an Demo to show how `DATETIME` got handled by `golang`, `gorm` and `mysql`

[klesh/gorm-mysql-timezone-demo](https://github.com/klesh/gorm-mysql-timezone-demo)

### Solution

1. Add `loc` setting to `DB_URI` to specify what timezone should mysql use to store `DATETIME` values.
2. `loc` is read from `TZ` Environment Variable, because this variable sets the default timezone for current golang process( dictate the timezone of `time.Now()`).
3. If `TZ` was omitted, `loc` should fall back to `Local` or it would be set to `UTC` by `gorm`, which is not so friendly when users try to browse the database directly.
4. Change all data type of `DATETIME` fields from string to `DATETIME`

## Q&A

- What if the server had a default timezone which was not what I wanted?
  Set the `TZ` Environment Variable before launching devlake
- Do I need to set `TZ` on my local machine?
  Normally, No! The timezone on your local machine is what your preferred, NORMALLY.
- My database is having all my date-time values wrong, like couple hours off.
  Delete database, Set the `TZ` Environment Variable and launch devlake to re-pull all the data
- Can I move the whole deployment to another server?
  Yes, as long as your TZ settings are consistence.

**Comment by klesh** [comment]
@joncodo @kevin-kline  Please take a look and leave your comment.

**Comment by liangjingyang** [comment]
I think we should use `DATETIME` data type in MySQL, and lock the `loc` on `DB_URI` to `UTC`.

**Comment by klesh** [comment]
@liangjingyang Nice, but why lock to `UTC`? This makes sense for most backend projects or data analysis systems, but for devlake, we assume user will use it for his team(s), I think a default `Local` setting makes more sense. Because many queries calculate on Natrual Month/Week basis...

**Comment by liangjingyang** [comment]
Lock `loc` to `UTC` can avoid data confusing by changing timezone configuration by mistake, because `gorm` will convert `time.Time` into `loc` timezone. It's also convenient to use `UNIX_TIMESTAMP/FROM_UNIXTIME ` functions, which also used by Grafana `$__time(dateColumn)`.

I think we could have 3 level timezone settings.

1. Locked `UTC` for `loc`. It's for collector or something else to insert data.
2. Timezone from config file of Lake. It's for enricher to aggregate data, and for user as default timezone. (Use `CONVERT_TZ` to convert timezone)
3. Timezone from MySQL read only session (session.time_zone). It's for some component like Grafana to query data.

**Comment by klesh** [comment]
I see, that's a great idea, I'd like the same thing. But unfortunately, it's impossible, mysql doesn't store timezone information on DATETIME values, nor does it convert the timezone dynamically session-wise for DATETIMEs.

> The session time zone setting **does not** affect values displayed by functions such as UTC_TIMESTAMP() or values in DATE, TIME, or DATETIME columns. Nor are values in those data types stored in UTC; the time zone applies for them only when converting from TIMESTAMP values. If you want locale-specific arithmetic for DATE, TIME, or DATETIME values, convert them to UTC, perform the arithmetic, and then convert back.
> https://dev.mysql.com/doc/refman/8.0/en/time-zone-support.html

I'm totally agree with you if it's only us querying the database,
But we are facing a situation that user may and most likely to query data directly on database, without timezone auto conversion, they will have to use `convert_tz` and pass in timezone he wanted, maybe we could have a Variable for that, then we have to worry about the synchronization between grafana and devlake, that's tedious and error prone

For example, we have many queries like this  from [grafana](http://localhost:3002/d/ZfaFQeM7k/delivery-velocity?orgId=1&editPanel=96&from=1622727795120&to=1622799621626)
```sql
with _requirement as(
  select 
    ji.id,
    DATE_ADD(date(resolution_date), INTERVAL -$interval(date(resolution_date))+1 DAY) as time,
    std_story_point
  from jira_issues ji
    join jira_board_issues jbi on ji.id = jbi.issue_id
  where 
    std_type = 'Requirement'
    and std_story_point != 0
    and $__timeFilter(created)
    and jbi.board_id = $board_id
)

select 
  timestamp(time) as time,
  avg(std_story_point) as average_story_point
from _requirement
group by 1
order by 1 asc
```
If we fixed the Timezone to UTC in database, each and every DATETIME feilds (resolution_date, created) must be wrapped within `convert_tz` function.

**Comment by liangjingyang** [comment]
Yes, the `Datetime` type doesn't care the seesion timezone, but `UNIX_TIMESTAMP` care.

![image](https://user-images.githubusercontent.com/2056705/134450448-02e1ab32-7095-4f59-811b-c54df6864b70.png)

![image](https://user-images.githubusercontent.com/2056705/134450470-c5ecb188-d7ff-48e5-b0d0-03c27c572697.png)


I'm not sure I understand the `average_story_point` SQL right(especially `-$interval(date(resolution_date))+1`). I assume that it calculates the average story point group by N days. If it's right, I think it looks like the thing that [`$__timeGroup(dateColumn,'5m')` do in Grafana](https://grafana.com/docs/grafana/latest/datasources/mysql).

And Grafana can also [change timezone at multiple levels](https://grafana.com/docs/grafana/latest/administration/preferences/change-grafana-timezone/). 

I guess, we don't need `convert_tz` to handle `Datetime` type on `UTC` timezone when using Grafana. The time macros and multiple level timezone settings in Grafana can resolve our concern.

But all above are just what I find from Grafana and MySQL websites, I never tested them. Please correct me if there is something wrong or they can not work, thanks.

**Comment by klesh** [comment]
Yes, UNIX_TIMESTAMP can be converted automatically, but it has a very limited time range up to year 2038. (17 years from now)
We might not need that long, but I certainly hope that we do... 
Your solution works if we opt for TIMESTAMP, but I vote against it.

Let's go back to DATETIME solution,
The problem wasn't about **time period**, it's about **timezone**
If we didn't use `convert_tz` on UTC-fixed DATETIME field, we end up with a UTC DATE, not a CST DATE, they are 8 hours off.

**Comment by klesh** [comment]
>I guess, we don't need convert_tz to handle Datetime type on UTC timezone when using Grafana. The time macros and multiple level timezone settings in Grafana can resolve our concern.
Yes, I agree, it's ok if only us would ever write those queries, we can enforce the rule of using time macros everywhere. But it's hard and unwise to enforce all users to do so.

For example, we assumed that user might use our app to collect data from JIRA only, and makes some charts he wanted, or, even worse, he wants to query the database directly from his own system( no time macro anymore). Say he would like to count Number of New Requirements per day, and his team located in China:
```sql
SELECT DATE_FORMAT(created, '%Y-%m-%d') as date, COUNT(*) as num_of_req
FROM issues
WHERE type = 'REQUIREMENT'
GROUP BY DATE_FORMAT(created, '%Y-%m-%d')
```
If someone created an issue at '2021-01-19 07:00 CST',  then this issue would be grouped into `2021-01-18` because of `created` being UTC.

**Comment by liangjingyang** [comment]
OK，I agree with your solution and Q&A in description of this issue. I think it is indeed easier to understand to use a unified timezone in Lake.

**Comment by klesh** [comment]
@joncodo @kevin-kline Please work on the issue on iter2
